File: python/FBT.py
#!/usr/bin/env python

###############################################################################
#                                                                             #
#                Fast Bessel Transform (FBT) for TMDs                         #
#     Zhongbo Kang, Alexei Prokudin, Nobuo Sato, John Terry                   #
#                   Please cite ArXiv:1906.05949                              #
#                      h is spacing parameter                                 #
#                  N is number of function calls                              #
#                  nu is Bessel function order                                #
#                                                                             #
###############################################################################

import numpy as np
from scipy.special import jv, jn_zeros, yv
from scipy.optimize import fsolve,minimize_scalar

class FBT:

    def __init__(self,nu=0.0):
        """ Constructor, sets nu of Jnu()                          """
        self.nu = nu
        self.setup()
        
    def setup(self,N = 100):
        self.jn_zeros0 = jn_zeros(self.nu,N)
        self.xi = self.jn_zeros0/np.pi
        Jp1=jv(self.nu+1,np.pi*self.xi)
        self.w=yv(self.nu, np.pi * self.xi) / Jp1
        
    def get_psi(self, t):
        return t*np.tanh(np.pi/2*np.sinh(t))
     
    def get_psip(self, t):
        return np.pi*t*(-np.tanh(np.pi*np.sinh(t)/2)**2 + 1)*np.cosh(t)/2 + np.tanh(np.pi*np.sinh(t)/2)

    def _ogatat(self,f,h,N,nu):
        """Transformed Ogata quadrature sum. Equation 8 in the reference."""
        N = int(N)
        if N > 100:
           self.setup(N)
        knots=np.pi/h*self.get_psi(h*self.xi[:N])
        Jnu=jv(nu,knots)
        psip=self.get_psip(h*self.xi[:N])
        # AP 5/3/23 This line does not when the functions does not accept a list as the argument
        #F=f(knots)
        # the solution:
        F=[f(knot) for knot in knots]
        psip[np.isnan(psip)]=1.0
        val=0.5*np.sum(self.w[:N]*F*Jnu*psip)
        return val

    def _ogatau(self,f,h,N,nu):
        """Untransformed Ogata quadrature sum. Equation 7 in the reference."""
        if N > 100:
           self.setup(N)
        knots = self.xi[:N]*h
        g=lambda x: f(x)*jv(nu,x)
        # AP 5/3/23 This line does not when the functions does not accept a list as the argument
        #F=g(knots)
        # the solution:
        F=[g(knot) for knot in knots]
        val=h*np.sum(self.w[:N]*F)/2./np.pi
        return val

    def _get_hu(self,f,q,Q):
        """Determines the untransformed hu by maximizing contribution to first node. Equation 11 in ref."""
        zero1 = self.jn_zeros0[0]
        h = lambda x: -abs(x*f(x/q))
        """Use brent method to maximize."""
        hu = minimize_scalar(h, bracket=None, bounds=(Q/10,10*Q), args=(), method='brent', tol=0.01, options=None).x/zero1*np.pi
        if hu>2.:
            hu = 2.
        return hu

    def _get_ht(self,hu,N):
        """Determine transformed ht from untransformed hu. Equation 13 in ref."""
        zeroN = self.jn_zeros0[int(N-1)]
        ht = np.pi/zeroN*np.arcsinh(2/np.pi*np.arctanh(hu/np.pi))
        return ht


    def fbt(self,g,q,N=10,Q=10.,option=0):
        """ Transformed optimized Ogata of a function f.                              """
        """            /Infty                                                         """
        """  result = |  d x g(x) J_nu(q*x)                                           """
        """           /0                                                              """
        """ Parameters                                                                """
        """ g: function of a single argument that has a single maximum in [0,Infinity)"""
        """ q: double precision:                                                      """
        """    the position where Fourier transform to be avaluated                   """
        """ N: int optional:                                                          """
        """    number of nodes to use                                                 """
        """ Q: double optional:                                                       """
        """    user's guess of the position of the maximum of function f              """
        """ option: int optional:                                                     """
        """    0 transformed Ogata  optimized h                                       """
        """    1 untransformed Ogata  optimized h                                     """
        """    2 transformed Ogata   with default h = 0.05                          """
        nu = self.nu
        f = lambda x: g(x/q)/q
        if option == 0:   # Transformed Ogata optimized h
            hu = self._get_hu(g,q,Q)
            ht = self._get_ht(hu,N)
            result = self._ogatat(f,ht,N,nu)
        elif option == 1: # Untransformed Ogata optimized h
            hu = self._get_hu(g,q,Q)
            result=self._ogatau(f,hu,N,nu)
        elif option == 2: # Transformed Ogata with h = 0.05
            hu = 0.05
            result=self._ogatat(f,hu,N,nu)
        return result

    def fbterror(self,g,q,N=10,Q=10.,option=0):
        """Transformed optimized Ogata error estimate."""
        N1 = N
        resultN =  self.fbt(g,q,N1,Q,option);
        N2 = 2*N
        result2N =  self.fbt(g,q,N2,Q,option);
        return np.abs((result2N-resultN)/result2N)

    def fbt_findN(self,g,q,err,Q=10,option=0):
        _N = 2
        _Q= Q
        _option = option
        W1 = self.fbt(g,q,N=  _N,Q=_Q,option=_option)
        W2 = self.fbt(g,q,N=2*_N,Q=_Q,option=_option)
        while abs(W1-W2)/W2>err:
            _N = 2*_N
            W1 = self.fbt(g,q,N=  _N,Q=_Q,option=_option)
            W2 = self.fbt(g,q,N=2*_N,Q=_Q,option=_option)
        return _N

File: cpp/FBT.cpp
//###############################################################################
//#                                                                             #
//#                Fast Bessel Transform (FBT) for TMDs                         #
//#     Zhongbo Kang, Alexei Prokudin, Nobuo Sato, John Terry                   #
//#                   Please cite ArXiv:1906.05949                              #
//#                  N is number of function calls                              #
//#                  nu is Bessel function order                                #
//###############################################################################
#define _USE_MATH_DEFINES // using M_PI for pi

#include <cmath> // abs
#include <math.h>
#include <stdio.h>
#include <iostream>
#include "FBT.h"

//acknowledgement
void FBT::acknowledgement(){
    std::cout << "###############################################################################" << std::endl;
    std::cout << "#                                                                             #" << std::endl;
    std::cout << "#                Fast Bessel Transform (FBT) for TMDs                         #" << std::endl;
    std::cout << "#     Zhongbo Kang, Alexei Prokudin, Nobuo Sato, John Terry                   #" << std::endl;
    std::cout << "#                   Please cite Kang:2019ctl                                  #" << std::endl;
    std::cout << "#                  N is number of function calls                              #" << std::endl;
    std::cout << "#                  nu is Bessel function order                                #" << std::endl;
    std::cout << "#                                                                             #" << std::endl;
    std::cout << "###############################################################################" << std::endl;
};

// Deconstructor
FBT::~FBT(){
  //jn_zeros0.~vector<double>();
};

// Constructor
FBT::FBT(double _nu, int _N, double _Q){
  if( _nu >= 0.){
    this->nu     = _nu;
  } else {
    std::cerr << " The value of nu = " << _nu << " is not supported." << std::endl;
    std::cerr << " Falling back to default  nu = " << FBT::nu_def << std::endl;
    this->nu     = FBT::nu_def;
  }

  if( _N >= 1){
    this->N     = _N;
  } else {
    std::cerr << " The value of N = " << _N << " is not supported." << std::endl;
    std::cerr << " Falling back to default  N = "  << FBT::N_def <<std::endl;
    this->N     = FBT::N_def;
  }


  if( _Q > 0){
    this->Q     = _Q;
  } else {
    std::cerr << " The value of Q = " << _Q << " is not supported." << std::endl;
    std::cerr << " Falling back to default  Q = "  << FBT::Q_def <<std::endl;
    this->Q     = FBT::Q_def;
  }

  // Sets maximum number of nodes to about 2^15:
  const int maxN = 32769;

  //Imports zeros of the Bessel function. Initializing this way speeds up calls
  try
  {
    boost::math::cyl_bessel_j_zero(this->nu, 1, maxN, std::back_inserter(jn_zeros0));
  }
  catch (std::exception& ex)
  {
    std::cout << "Thrown exception " << ex.what() << std::endl;
  }

  acknowledgement();

};

double get_psi( double t){
  return ( t )*tanh( M_PI/2.* sinh( t ) );
};

double get_psip( double t){
  return M_PI*t*( -pow(tanh( M_PI*sinh(t)/2.),2) + 1.)*cosh(t)/2. + tanh(M_PI*sinh(t)/2.);
};


double f_for_ogata(double x, double (*g)(double,void*), void* data, double q){
  return g(x/q,data)/q;
};

//Transformed Ogata quadrature sum. Equation ? in the reference.
double FBT::ogatat(double (*f)(double,void*), void* data, double q, double h){
  double nu = this->nu;
  int N = this->N;

  std::vector<double> zeros;
  zeros.resize(N); // Allocate N elements and copy from jn_zeros

  std::vector<double> xi;
  xi.resize(N);


  std::vector<double> Jp1;
  Jp1.resize(N);

  std::vector<double> w;
  w.resize(N);

  std::vector<double> knots;
  knots.resize(N);

  std::vector<double> Jnu;
  Jnu.resize(N);

  std::vector<double> psip;
  psip.resize(N);


  std::vector<double> F;
  F.resize(N);

  double val = 0;

  try{
    for (size_t i = 0; i < (unsigned)N; i++) {
      zeros[i] = jn_zeros0[i];
      xi[i] = zeros[i]/M_PI;
      Jp1[i] = boost::math::cyl_bessel_j(nu+1.,M_PI*xi[i]); //The functions cyl_bessel_j and cyl_neumann return the result of the Bessel functions of the first and second kinds respectively
      w[i] = boost::math::cyl_neumann(nu,M_PI*xi[i])/Jp1[i];
      knots[i] = M_PI/h*get_psi( h*xi[i] );
      Jnu[i] = boost::math::cyl_bessel_j(nu,knots[i]);
      double temp =  get_psip(h*xi[i]);

      if( isnan(temp) )
      {
        psip[i] = 1.;
      }
      else
      {
        psip[i] = temp;
      };
      F[i] = f_for_ogata(knots[i], f, data, q);
      val += M_PI*w[i]*F[i]*Jnu[i]*psip[i];


    }
  }
  catch (std::exception& ex)
  {
    std::cout << "Thrown exception " << ex.what() << std::endl;
  }

  return val;
};

//"""Untransformed Ogata quadrature sum. Equation ? in the reference."""
double FBT::ogatau(double (*f)(double,void*), void* data, double q, double h){
  double nu = this->nu;
  int N = this->N;

  std::vector<double> zeros;
  zeros.resize(N); // Allocate N elements and copy from jn_zeros

  std::vector<double> xi;
  xi.resize(N);

  std::vector<double> Jp1;
  Jp1.resize(N);

  std::vector<double> w;
  w.resize(N);

  std::vector<double> knots;
  knots.resize(N);

  std::vector<double> F;
  F.resize(N);

  double val = 0;

  try
  {
    for (size_t i = 0; i < (unsigned)N; i++) {
      zeros[i] = jn_zeros0[i];
      xi[i] = zeros[i]/M_PI;
      Jp1[i]=boost::math::cyl_bessel_j(nu+1,M_PI*xi[i]);
      w[i]=boost::math::cyl_neumann(nu,M_PI*xi[i])/Jp1[i];
      knots[i] = xi[i]*h;
      F[i]=f_for_ogata(knots[i], f, data, q)*boost::math::cyl_bessel_j(nu,knots[i]);
      val+=h*w[i]*F[i];
    }
  }
  catch(std::exception& ex)
  {
    std::cout << "Thrown exception " << ex.what() << std::endl;
  }

  return val;
};

double f_for_get_hu(double x, double (*g)(double,void*), void* data, double q){
   return -abs(x*g(x/q, data));
};

//"""Determines the untransformed hu by maximizing contribution to first node. Equation ? in ref."""
double FBT::get_hu(double (*f)(double,void*), void* data, double q){
  double Q = this->Q;

  double zero1 = jn_zeros0[0];
  const int double_bits = std::numeric_limits<double>::digits;
  std::pair<double, double> r = boost::math::tools::brent_find_minima(std::bind(f_for_get_hu, std::placeholders::_1, f, data, q), Q/10., 10.*Q, double_bits);

  double hu = r.first/zero1;
  if(hu >= 3.){
    hu = 3.;
    std::cerr<< "Warning: Number of nodes is too small N = " << this->N << std::endl;
  }

  return hu;
};

// double f_for_get_ht(double x, double hu, double zeroN){
//   return hu-M_PI*tanh(M_PI/2.*sinh(x*zeroN/M_PI));
// };

//"""Determine transformed ht from untransformed hu. Equation ? in ref."""
double FBT::get_ht(double hu){
  int N = this->N;

  double zeroN = double(jn_zeros0[N-1]);

  return M_PI/zeroN*asinh(2./M_PI*atanh(hu/M_PI));


};

//"""Untransformed optimized Ogata."""
double FBT::fbtu(double (*g)(double,void*), void* data, double q){
  /* Numerical computation of F(q)=int(f(x)*Jn(x*q)) Bessel transform
		[in]g     - integrand
		[in]data  - pointer on user-defined data which will
		[in]data  - pointer on user-defined data which will
					be passed to g every time it called (as third parameter).
		[in]q - evaluation point

	return:
			-computed integral value
	*/
  double hu = get_hu(g,data,q);
  return ogatau(g,data,q,hu);
};

//"""Transformed optimized Ogata."""
double FBT::fbt(double (*g)(double,void*), void* data, double q){
  /* Numerical computation of F(q)=int(f(x)*Jn(x*q)) Bessel transform
		[in]g     - integrand
		[in]data  - pointer on user-defined data which will
		[in]data  - pointer on user-defined data which will
					be passed to g every time it called (as third parameter).
		[in]q - evaluation point

	return:
			-computed integral value
	*/
  double hu = get_hu(g,data,q);
  double ht = get_ht(hu);
  return ogatat(g,data,q,ht);
};
